File: src/remi/memory/daemon.py
"""Memory daemon — consume queued transcripts and run maintenance agent.

Watches ~/.remi/queue/ for new .jsonl files, processes them through an LLM
to extract memory-worthy information, and patches memory files.
"""

from __future__ import annotations

import asyncio
import json
import logging
import time
from pathlib import Path

from remi.memory.maintenance import build_maintenance_prompt
from remi.memory.store import MemoryStore

logger = logging.getLogger(__name__)

LOCK_FILE = ".maintenance.lock"
LOCK_TIMEOUT = 60  # seconds


class MemoryDaemon:
    """Watch queue directory and process transcripts."""

    def __init__(
        self,
        store: MemoryStore,
        queue_dir: Path | None = None,
        poll_interval: float = 10.0,
    ) -> None:
        self.store = store
        self.queue_dir = queue_dir or Path.home() / ".remi" / "queue"
        self.queue_dir.mkdir(parents=True, exist_ok=True)
        (self.queue_dir / "processed").mkdir(exist_ok=True)
        self.poll_interval = poll_interval

    async def run(self, shutdown_event: asyncio.Event | None = None) -> None:
        """Main loop: poll queue directory for new items."""
        logger.info("MemoryDaemon started, watching %s", self.queue_dir)
        while True:
            if shutdown_event and shutdown_event.is_set():
                break
            try:
                await self._process_queue()
            except Exception as e:
                logger.error("Queue processing error: %s", e)
            if shutdown_event:
                try:
                    await asyncio.wait_for(shutdown_event.wait(), timeout=self.poll_interval)
                    break
                except asyncio.TimeoutError:
                    pass
            else:
                await asyncio.sleep(self.poll_interval)
        logger.info("MemoryDaemon stopped.")

    async def _process_queue(self) -> None:
        """Process all pending .jsonl files in the queue."""
        lock_path = self.store.root / LOCK_FILE
        if lock_path.exists():
            mtime = lock_path.stat().st_mtime
            if time.time() - mtime < LOCK_TIMEOUT:
                logger.debug("Maintenance lock held, skipping")
                return
            # Stale lock
            lock_path.unlink()

        pending = sorted(self.queue_dir.glob("*.jsonl"))
        if not pending:
            return

        # Acquire lock
        lock_path.write_text(str(time.time()), encoding="utf-8")
        try:
            for jsonl_file in pending:
                await self._process_file(jsonl_file)
        finally:
            lock_path.unlink(missing_ok=True)

    async def _process_file(self, jsonl_file: Path) -> None:
        """Process a single queued transcript file."""
        try:
            data = json.loads(jsonl_file.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError) as e:
            logger.error("Failed to read %s: %s", jsonl_file, e)
            return

        content_hash = data.get("hash", "")

        # Idempotency check
        processed_file = self.queue_dir / ".processed"
        if processed_file.exists():
            processed = processed_file.read_text(encoding="utf-8").splitlines()
            if content_hash in processed:
                logger.debug("Already processed %s, skipping", content_hash)
                jsonl_file.unlink()
                return

        transcript = data.get("transcript", "")
        if not transcript.strip():
            jsonl_file.unlink()
            return

        logger.info("Processing queued transcript: %s", jsonl_file.name)

        # Build prompt and process (placeholder for LLM integration)
        prompt = build_maintenance_prompt(
            cwd=None,
            summary="",
            recent_turns=transcript[:5000],
            memory_structure=self._describe_memory_structure(),
        )

        # TODO: call LLM with prompt, parse response, execute actions
        # For now, just log and move to processed
        logger.info("Maintenance prompt built (%d chars), LLM call pending", len(prompt))

        # Record as processed
        with processed_file.open("a", encoding="utf-8") as f:
            f.write(f"{content_hash}\n")

        # Move to processed directory
        dest = self.queue_dir / "processed" / jsonl_file.name
        jsonl_file.rename(dest)
        logger.info("Moved %s to processed/", jsonl_file.name)

    def _describe_memory_structure(self) -> str:
        """Describe current memory layout for the maintenance prompt."""
        lines = []
        entities_dir = self.store.root / "entities"
        if entities_dir.is_dir():
            for type_dir in sorted(entities_dir.iterdir()):
                if type_dir.is_dir():
                    count = len(list(type_dir.glob("*.md")))
                    lines.append(f"  entities/{type_dir.name}/: {count} files")

        daily_dir = self.store.root / "daily"
        if daily_dir.is_dir():
            count = len(list(daily_dir.glob("*.md")))
            lines.append(f"  daily/: {count} files")

        return "\n".join(lines) if lines else "  (empty)"

    def cleanup_processed(self, keep_days: int = 30) -> int:
        """Remove processed files older than keep_days."""
        from datetime import timedelta

        cutoff = time.time() - timedelta(days=keep_days).total_seconds()
        removed = 0
        processed_dir = self.queue_dir / "processed"
        if processed_dir.is_dir():
            for f in processed_dir.glob("*.jsonl"):
                if f.stat().st_mtime < cutoff:
                    f.unlink()
                    removed += 1
        return removed

File: src/memory/daemon.ts
/**
 * Memory daemon — consume queued transcripts and run maintenance agent.
 *
 * Watches ~/.remi/queue/ for new .jsonl files, processes them through an LLM
 * to extract memory-worthy information, and patches memory files.
 */

import {
  existsSync,
  mkdirSync,
  readFileSync,
  writeFileSync,
  appendFileSync,
  unlinkSync,
  readdirSync,
  statSync,
  renameSync,
} from "node:fs";
import { join } from "node:path";
import { homedir } from "node:os";
import { buildMaintenancePrompt } from "./maintenance.js";
import type { MemoryStore } from "./store.js";
import { createLogger } from "../logger.js";

const log = createLogger("mem-daemon");

const LOCK_FILE = ".maintenance.lock";
const LOCK_TIMEOUT = 60; // seconds

export class MemoryDaemon {
  store: MemoryStore;
  queueDir: string;
  pollInterval: number;

  constructor(
    store: MemoryStore,
    options?: { queueDir?: string; pollInterval?: number },
  ) {
    this.store = store;
    this.queueDir = options?.queueDir ?? join(homedir(), ".remi", "queue");
    this.pollInterval = options?.pollInterval ?? 10.0;

    if (!existsSync(this.queueDir)) {
      mkdirSync(this.queueDir, { recursive: true });
    }
    const processedDir = join(this.queueDir, "processed");
    if (!existsSync(processedDir)) {
      mkdirSync(processedDir, { recursive: true });
    }
  }

  async run(shutdownSignal?: AbortSignal): Promise<void> {
    log.info(`MemoryDaemon started, watching ${this.queueDir}`);

    while (!shutdownSignal?.aborted) {
      try {
        await this._processQueue();
      } catch (e) {
        log.error("Queue processing error:", e);
      }

      // Wait for poll interval or shutdown
      await new Promise<void>((resolve) => {
        const timer = setTimeout(resolve, this.pollInterval * 1000);
        if (shutdownSignal) {
          shutdownSignal.addEventListener("abort", () => {
            clearTimeout(timer);
            resolve();
          }, { once: true });
        }
      });
    }

    log.info("MemoryDaemon stopped.");
  }

  private async _processQueue(): Promise<void> {
    const lockPath = join(this.store.root, LOCK_FILE);
    if (existsSync(lockPath)) {
      const mtime = statSync(lockPath).mtimeMs / 1000;
      if (Date.now() / 1000 - mtime < LOCK_TIMEOUT) {
        return; // Lock held
      }
      // Stale lock
      unlinkSync(lockPath);
    }

    const pending = readdirSync(this.queueDir)
      .filter((f) => f.endsWith(".jsonl"))
      .sort();

    if (pending.length === 0) return;

    // Acquire lock
    writeFileSync(lockPath, String(Date.now() / 1000), "utf-8");
    try {
      for (const file of pending) {
        await this._processFile(join(this.queueDir, file));
      }
    } finally {
      if (existsSync(lockPath)) {
        unlinkSync(lockPath);
      }
    }
  }

  private async _processFile(jsonlFile: string): Promise<void> {
    let data: Record<string, unknown>;
    try {
      data = JSON.parse(readFileSync(jsonlFile, "utf-8"));
    } catch (e) {
      log.error(`Failed to read ${jsonlFile}:`, e);
      return;
    }

    const contentHash = (data.hash as string) ?? "";

    // Idempotency check
    const processedFile = join(this.queueDir, ".processed");
    if (existsSync(processedFile)) {
      const processed = readFileSync(processedFile, "utf-8").split("\n");
      if (processed.includes(contentHash)) {
        unlinkSync(jsonlFile);
        return;
      }
    }

    const transcript = (data.transcript as string) ?? "";
    if (!transcript.trim()) {
      unlinkSync(jsonlFile);
      return;
    }

    log.info(`Processing queued transcript: ${jsonlFile.split("/").pop()}`);

    // Build prompt (placeholder for LLM integration)
    const prompt = buildMaintenancePrompt(
      null,
      "",
      transcript.slice(0, 5000),
      this._describeMemoryStructure(),
    );

    // TODO: call LLM with prompt, parse response, execute actions
    log.info(`Maintenance prompt built (${prompt.length} chars), LLM call pending`);

    // Record as processed
    appendFileSync(processedFile, `${contentHash}\n`, "utf-8");

    // Move to processed directory
    const dest = join(this.queueDir, "processed", jsonlFile.split("/").pop()!);
    renameSync(jsonlFile, dest);
  }

  _describeMemoryStructure(): string {
    const lines: string[] = [];
    const entitiesDir = join(this.store.root, "entities");
    if (existsSync(entitiesDir)) {
      for (const entry of readdirSync(entitiesDir, { withFileTypes: true })) {
        if (entry.isDirectory()) {
          const count = readdirSync(join(entitiesDir, entry.name)).filter((f) =>
            f.endsWith(".md"),
          ).length;
          lines.push(`  entities/${entry.name}/: ${count} files`);
        }
      }
    }

    const dailyDir = join(this.store.root, "daily");
    if (existsSync(dailyDir)) {
      const count = readdirSync(dailyDir).filter((f) => f.endsWith(".md")).length;
      lines.push(`  daily/: ${count} files`);
    }

    return lines.length > 0 ? lines.join("\n") : "  (empty)";
  }

  cleanupProcessed(keepDays: number = 30): number {
    const cutoff = Date.now() - keepDays * 24 * 60 * 60 * 1000;
    let removed = 0;
    const processedDir = join(this.queueDir, "processed");
    if (!existsSync(processedDir)) return 0;

    for (const file of readdirSync(processedDir)) {
      if (!file.endsWith(".jsonl")) continue;
      const fullPath = join(processedDir, file);
      if (statSync(fullPath).mtimeMs < cutoff) {
        unlinkSync(fullPath);
        removed++;
      }
    }
    return removed;
  }
}
